### PR TITLE
[SPARK-21687][SQL] Spark SQL should set createTime for Hive partition

### PR DESCRIPTION
**Pull request by debugger87**
## What changes were proposed in this pull request?

Set createTime for every hive partition created in Spark SQL, which could be used to manage data lifecycle in Hive warehouse. We found  that almost every partition modified by spark sql has not been set createTime.

```
mysql> select * from partitions where create_time=0 limit 1\G;
*************************** 1. row ***************************
         PART_ID: 1028584
     CREATE_TIME: 0
LAST_ACCESS_TIME: 1502203611
       PART_NAME: date=20170130
           SD_ID: 1543605
          TBL_ID: 211605
  LINK_TARGET_ID: NULL
1 row in set (0.27 sec)
```

## How was this patch tested?
 N/A

### CODE REVIEW

**Comment by debugger87** [comment]
@cloud-fan could you please help me to review this PR?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -986,6 +986,7 @@ private[hive] object HiveClientImpl {
     tpart.setTableName(ht.getTableName)
     tpart.setValues(partValues.asJava)
     tpart.setSd(storageDesc)
+    tpart.setCreateTime((System.currentTimeMillis() / 1000).toInt)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: This is to Hive, how about from Hive?

Comment (debugger87) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: We just need to use API of hive_metastore like `get_partition` to fetch related information of Hive Partition.

* Partition in hive_metastore.thrift

```
struct Partition {
  1: list<string> values // string value is converted to appropriate partition key type
  2: string       dbName,
  3: string       tableName,
  4: i32          createTime,
  5: i32          lastAccessTime,
  6: StorageDescriptor   sd,
  7: map<string, string> parameters,
  8: optional PrincipalPrivilegeSet privileges
}
```

* get_partition in hive_metastore.thrit
```
Partition get_partition(1:string db_name, 2:string tbl_name, 3:list<string> part_vals)
                       throws(1:MetaException o1, 2:NoSuchObjectException o2)
```
We will set TTL as kv into parameters of `Table` and fetch `createTime` from `Partition` to decide if we can drop a partition in Hive.

**Comment by rxin** [comment]
We should put this in the catalog, shouldn't we?

**Comment by cloud-fan** [comment]
+1, similar to `CatalogTable.createTime`, we should have a `CatalogTablePartition.createTime`

**Comment by debugger87** [comment]
My changes is not enough to support `createTime` in CatalogTablePartition, I will check and re-commit again.

**Comment by debugger87** [comment]
@cloud-fan Have a look at this PR again?

**Comment by debugger87** [comment]
@cloud-fan Look at this PR again? I just put `createTime` into CatalogTablePartion.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
Comment (debugger87) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @gatorsmile sorry for my misunderstanding of your point. `toHivePartition` and `fromHivePartition` should be symmetric, and I change my implementation. could you please review it again?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -97,7 +97,9 @@ object CatalogStorageFormat {
 case class CatalogTablePartition(
     spec: CatalogTypes.TablePartitionSpec,
     storage: CatalogStorageFormat,
-    parameters: Map[String, String] = Map.empty) {
+    parameters: Map[String, String] = Map.empty,
+    createTime: Long = System.currentTimeMillis,
+    lastAccessTime: Long = -1) {
 
   def toLinkedHashMap: mutable.LinkedHashMap[String, String] = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: You also need to add it to this map for display

**Comment by gatorsmile** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 2 above)
Comment (debugger87) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: @gatorsmile  Thanks for your reminding, i will add it.

**Comment by debugger87** [comment]
`createTime` is set by HiveMetaStore#initializeAddedPartition

```
private void initializeAddedPartition(Table tbl, PartitionIterator part, boolean madeDir) throws MetaException {
      if(HiveConf.getBoolVar(this.hiveConf, ConfVars.HIVESTATSAUTOGATHER) && !MetaStoreUtils.isView(tbl)) {
        MetaStoreUtils.updatePartitionStatsFast(part, this.wh, madeDir, false);
      }

      long time = System.currentTimeMillis() / 1000L;
      part.setCreateTime((long)((int)time));
      if(part.getParameters() == null || part.getParameters().get("transient_lastDdlTime") == null) {
        part.putToParameters("transient_lastDdlTime", Long.toString(time));
      }
      // ignore code lines
}
```

This PR should be closed and we will check the reason why createTime is zero for partitions created by spark sql again.

**Comment by cxzl25** [comment]
**Modify the partition will lose createTime.**
Reading the hive partitions ignores createTime when converting the CatalogTablePartition, it will also be lost when modifying partitions.
Calling this method SessionCatalog#alterPartitions will be lost createTime.
So can you reopen this pr? @debugger87 

```sql
CREATE  TABLE `tmp_test_partition`(
  `c1` string
)
PARTITIONED BY (
  `d` string
);
ALTER TABLE `tmp_test_partition_1` ADD PARTITION (d='1');
ALTER TABLE `tmp_test_partition_1` PARTITION (d='1') SET LOCATION 'xxx';
```

**Comment by debugger87** [comment]
@cxzl25 OK, reopen it

**Comment by cloud-fan** [comment]
ok to test

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -1019,6 +1021,8 @@ private[hive] object HiveClientImpl {
         compressed = apiPartition.getSd.isCompressed,
         properties = Option(apiPartition.getSd.getSerdeInfo.getParameters)
           .map(_.asScala.toMap).orNull),
+      createTime = apiPartition.getCreateTime.toLong * 1000,
+      lastAccessTime = apiPartition.getLastAccessTime.toLong * 1000)
```
Comment (cxzl25) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Add a comma to the end?

Comment (debugger87) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @cxzl25 yeah, it's my mistake, i will fix it

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -1019,6 +1021,8 @@ private[hive] object HiveClientImpl {
         compressed = apiPartition.getSd.isCompressed,
         properties = Option(apiPartition.getSd.getSerdeInfo.getParameters)
           .map(_.asScala.toMap).orNull),
+      createTime = apiPartition.getCreateTime.toLong * 1000,
+      lastAccessTime = apiPartition.getLastAccessTime.toLong * 1000,
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Can we use `DurationConversions` here?

**Comment by maropu** [comment]
Any update?

**Comment by debugger87** [comment]
@cxzl25  Any suggestions?

```
org.apache.spark.sql.streaming.FlatMapGroupsWithStateSuite.flatMapGroupsWithState - streaming with processing time timeout
```

**Comment by debugger87** [comment]
@cxzl25 I run `build/mvn  -Dtest=none -DwildcardSuites=org.apache.spark.sql.streaming.FlatMapGroupsWithStateSuite  test` in my local PC, output is as follows:

```
Run completed in 36 seconds, 46 milliseconds.
Total number of tests run: 73
Suites: completed 2, aborted 0
Tests: succeeded 73, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

**Comment by debugger87** [comment]
@cxzl25  @maropu any suggestions？

**Comment by debugger87** [comment]
@cloud-fan I found that if we follow code in CatalogTable.toLinkedHashMap, this patch will break so many test cases in org.apache.spark.sql.SQLQueryTestSuite (compare *.out files).

**Comment by cloud-fan** [comment]
can you check if hive DESC TABLE  output the create time of partitions? We can regenerate the output files for `SQLQueryTestSuite` if needed.

**Comment by debugger87** [comment]
@cloud-fan For Hive, `desc extended dbName.table` or `desc extended dbName.table partition (name=value)` will print `Detailed Table Information` or `Detailed Partition Information`. Of course, `createTime` will be in output console.

However, it's not needed to regenerate `SQLQueryTestSuite ` because of so many changes/diff generated.

**Comment by cloud-fan** [comment]
If hive outputs createTime for partitions, we should do it too, right?

**Comment by debugger87** [comment]
@cloud-fan totally agree with you, i will try to regenerate output files for `SQLQueryTestSuite`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 5]
```diff
@@ -93,12 +93,16 @@ object CatalogStorageFormat {
  * @param spec partition spec values indexed by column name
  * @param storage storage format of the partition
  * @param parameters some parameters for the partition
+ * @param createTime creation time of the partition
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: let's mention the time unit, i.e. in milliseconds.

Comment (debugger87) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: OK, it's the same as CatalogTable, in milliseconds. I fill fix this comment.

**Comment by debugger87** [comment]
@cloud-fan Any suggestions?

**Comment by gatorsmile** [comment]
LGTM Thanks! Merged to master.

**Pull request closed without merging**